### PR TITLE
OpInfo: Allow sample_inputs_func to be any iterable

### LINKED ISSUE #52486: OpInfo to support `sample_inputs_func` with just an Iterable output.
## 🚀 Feature
It would make things easier and more transparent/modular if OpInfo could support samplers which comply with the Iterable protocol.

## Motivation & Pitch

OpInfo does support a parameter `sample_inputs_func` which generates inputs for test cases. However, the `len` function is applied to the output of this function meaning that whichever output `sample_inputs_func` returns, it has to comply with the Sequence protocol. This comes with certain limitations:

* Encourages writing a one long sampler with either a huge tuple, or with appending to a list which does not look elegant, nor splits testing into independent parts.
* Does not allow mixing in generators or arbitrary iterables for that matter. That is a matter of taste, but using `yield ...` instead of `my_list.append(...)` seems to look better.
* If we are to split a sampler into subsamplers, we have to make sure that they all generate the SAME sequence such that we could concatenate it like, for example,
```python
def sampler_for_op(*args, **kwargs):
    return sampler1(*args, **kwargs) + sampler2(*args, **kwargs) + ...
    
    ...
    sample_inputs_func = sampler_for_op,
    ...
```
Or we could hack different type Sequences by using unpacking, but still it does not work with generators, unless converted to Sequences:
```python
def sampler_for_op(*args, **kwargs):
    return [*sampler1(*args, **kwargs), *sampler2(*args, **kwargs), ..., *list(generator_sampler(*args, **kwargs))]
```
If `sample_inputs_func` were Iterable, we could have simply chained arbitrary Iterable samplers like
```python
def sampler_for_op(*args, **kwargs):
    yield from sampler1(*args, **kwargs)
    yield from sampler2(*args, **kwargs)
    ...

    ...
    sample_inputs_func = sampler_for_op,
    ...
```
Not only does it make adding new samplers easier and more memory efficient, this way we could reuse same samplers for different operations by adding them to different chains.

Adding support for this feature should not break the existing code since Sequences are Iterable.

CC @mruberry 

cc @mruberry @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69257
* __->__ #69256

Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -4973,9 +4965,10 @@ def gen_inputs():
                     # Either A or B needs to have a gradient
                     if not grad_A and not grad_B:
                         continue
-                    A.requires_grad_(grad_A)
-                    B.requires_grad_(grad_B)
-                    yield SampleInput(A, args=(B,), kwargs=kwargs)
+                    yield SampleInput(
+                        A.clone().requires_grad_(grad_A),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4969]: Why is the clone needed here?

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -3103,3 +3103,10 @@ def measure() -> float:
         vals.append(measure())
     vals = sorted(vals)
     return mean(vals[2 : num - 2])
+
+
+def first_sample(self, samples):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:3108]: Add a comment documenting this function

**Review by mruberry — APPROVED** [review]
Thanks @peterbell10!

See two minor questions/requests inline

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32942008).

**Comment by Lezcano** [comment]
@nikitaved the day has come.

**Comment by nikitaved** [comment]
There is an outstanding PR: https://github.com/pytorch/pytorch/pull/53044. When compared to the one by @peterbell10 , it does not materialize the full sequence at once. Maybe it is better now since we have several wrappers which change the global random state without restoring it.

**Comment by peterbell10** [comment]
>When compared to the one by @peterbell10 , it does not materialize the full sequence at once.

I only call `list(samples)` in a few tests that iterate over the samples multiple times. There are still many test cases that iterate over the samples one at a time.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:5216]: @mruberry GitHub cuts this off in the diff, but it's inside a loop
```
for grad_A, grad_B in product((True, False), repeat=2):
  # Either A or B needs to have a gradient
  if not grad_A and not grad_B:
    continue
  A.requires_grad_(grad_A)
  B.requires_grad_(grad_B)
  yield SampleInput(A, args=(B,), kwargs=kwargs)
```

So previously the same tensor was returned in multiple `SampleInputs` simultaneously which causes this assert to fail
https://github.com/pytorch/pytorch/blob/fc8404b5bc7e9721aa93021cdc27de818df64d8d/test/test_ops.py#L251

File: torch/testing/_internal/common_utils.py (hunk 2 above)
Comment (peterbell10) [reply torch/testing/_internal/common_utils.py]: Done

**Comment by nikitaved** [comment]
> I only call `list(samples)` in a few tests that iterate over the samples multiple times. There are still many test cases that iterate over the samples one at a time.

I see, so there are tests which follow the processing like this `sample1 -> op(sample1) -> sample2 -> op(sample2) -> ...`? If so, we need to enable the fixture that is commented out in https://github.com/pytorch/pytorch/pull/69107#issuecomment-988095274, not not alter the global random state.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:5216]: This clone is required to make inputs independent yet with the same content.

**Comment by peterbell10** [comment]
I managed to get `freeze_rng_state` working with `test_composite_compliance`.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32942008).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32942008).

**Comment by peterbell10** [comment]
Master is failing because #69399 added a test that indexes into `sample_inputs`. #69697 will fix that test.

File: torch/testing/_internal/common_utils.py [hunk 3]
```diff
@@ -1123,13 +1124,21 @@ def set_rng_seed(seed):
...
 @contextlib.contextmanager
 def freeze_rng_state():
-    rng_state = torch.get_rng_state()
-    if torch.cuda.is_available():
-        cuda_rng_state = torch.cuda.get_rng_state()
-    yield
-    if torch.cuda.is_available():
-        torch.cuda.set_rng_state(cuda_rng_state)
-    torch.set_rng_state(rng_state)
+    # no_dispatch needed for test_composite_compliance
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/testing/_internal/common_utils.py:1127]: Hmm, so the implication here is that torch.get_rng_state and torch.set_rng_state are not composite compliant, right?

[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/testing/_internal/common_utils.py:1127]: No, the problem is that `Generator.set_rng_state()` doesn't support `__torch_dispatch__` which is used to implement the test.

**Pull request closed without merging**